### PR TITLE
Add support for color models to \colorbox

### PR DESCRIPTION
**Pull request by pkra**
Part of MathJax/MathJax#2871

### CODE REVIEW

**Comment by pkra** [comment]
Feel free to close & replace with another.

**Comment by dpvc** [comment]
Would you consider adding support for that in `\fcolorbox` as well?

**Comment by pkra** [comment]
> Would you consider adding support for that in \fcolorbox as well?

Happy to.

I wasn't sure what you had in mind with 

> Also, the `padded()` function should be exported.

on the issue thread.

**Comment by dpvc** [comment]
> I wasn't sure what you had in mind with
> > Also, the padded() function should be exported.
> 
> on the issue thread.

It means add "export" to this line:

https://github.com/mathjax/MathJax-src/blob/9f588aaf7ed8e7f38bfb636157bf29a94ee4804d/ts/input/tex/color/ColorMethods.ts#L40

**Comment by pkra** [comment]
> It means add "export" to this line:

Oh, padding, not padded :see_no_evil:. That makes sense then.

**Comment by dpvc** [comment]
OOPS!, my fault.

**Comment by pkra** [comment]
> OOPS!, my fault.

I feel like I should've been able to figure that one out :rofl: but thank you.

**Comment by pkra** [comment]
`\fcolorbox` supports four patterns:

```
{⟨frame color ⟩}{⟨background color ⟩}{⟨text⟩}
[⟨model-list⟩]{⟨frame spec-list⟩}{⟨background spec-list⟩}{⟨text⟩}
[⟨fr. model-list⟩]{⟨fr. spec-list⟩}[⟨backgr. model-list⟩]{⟨backgr. spec-list⟩}{⟨text⟩}
{⟨frame color ⟩}[⟨background model-list⟩]{⟨background spec-list⟩}{⟨text⟩}
```

Currently, MathJax supports pattern 1.
My commit tries to add support for pattern 2.

Let me know what you think.

File: ts/input/tex/color/ColorMethods.ts [hunk 1]
```diff
@@ -140,15 +139,16 @@ ColorMethods.ColorBox = function (parser: TexParser, name: string) {
  * @param {string} name The name of the control sequence.
  */
 ColorMethods.FColorBox = function (parser: TexParser, name: string) {
+  const model = parser.GetBrackets(name, "");
   const fname = parser.GetArgument(name);
   const cname = parser.GetArgument(name);
```
Comment (dpvc) [inline ts/input/tex/color/ColorMethods.ts:144]: I think the way to do this is
```suggestion
  const fmodel = parser.GetBrackets(name, 'named');
  const fname = parser.GetArgument(name);
  const cmodel = parser.GetBrackets(name, fmodel);
  const cname = parser.GetArgument(name);
```

and then use `fmodel` to get the color model for the frame and `cmodel` to get the model for the background color.  So you will need `frameModel` in addition to `colorModel` below.  (I haven't tested it, but this should be the right idea.)

File: ts/input/tex/color/ColorMethods.ts [hunk 2]
```diff
@@ -121,14 +121,13 @@ ColorMethods.DefineColor = function (parser: TexParser, name: string) {
  * @param {string} name The name of the control sequence.
  */
 ColorMethods.ColorBox = function (parser: TexParser, name: string) {
-  const cname = parser.GetArgument(name);
+  const model = parser.GetBrackets(name, "");
```
Comment (dpvc) [inline ts/input/tex/color/ColorMethods.ts:124]: Note that our coding style uses single quotation marks, not double, for string constants (tslint should flag this).  This occurs several times below as well.

**Review by dpvc — CHANGES_REQUESTED** [review]
Thanks for the update.  I've made a few comments about how to handle the `\fcolorbox` situation.

File: ts/input/tex/color/ColorMethods.ts (hunk 1 above)
Comment (pkra) [reply ts/input/tex/color/ColorMethods.ts:144]: Thanks! I'll try to find time this week.

**Comment by pkra** [comment]
Unexpected slot showed up so I tried to wrap this up.

I noticed that something like `\fcolorbox{127, 127, 127}[RGB]{155, 233, 168}{x}` would create ` style="border: 2px solid 127, 127, 127"` (whereas I'd expect an error of some kind).

I've been wondering what to do about tests.  I tried to spin up the v3 lab in mathjax-dev but failed (first, that eval bug, then systemjs stuff). I could get the scripts to work (without esm) but it's not clear to me what to do with them.

Let me know what you need from me here.

**Comment by pkra** [comment]
Thanks for your comments.

> Yeah, the named color model doesn't do any checking

Ok, I'll ignore that then (as an issue with getColor / getColorByName)

> @zorkow may be able to tell you about how to handle the tests.

Thanks!

**Comment by pkra** [comment]
@dpvc @zorkow was there anything else you need from me on this?

File: ts/input/tex/color/ColorMethods.ts [hunk 3]
```diff
@@ -121,14 +121,13 @@ ColorMethods.DefineColor = function (parser: TexParser, name: string) {
  * @param {string} name The name of the control sequence.
  */
 ColorMethods.ColorBox = function (parser: TexParser, name: string) {
-  const cname = parser.GetArgument(name);
+  const model = parser.GetBrackets(name, '');
```
Comment (dpvc) [inline ts/input/tex/color/ColorMethods.ts:124]: Does `''` need to be `'named'`, like you do in `FColorBox` below?

**Comment by dpvc** [comment]
I'm good with it, pending the one small change I just noted.  Can you change it from draft to a regular PR?

@zorkow, are you OK with the tests?

File: ts/input/tex/color/ColorMethods.ts (hunk 3 above)
Comment (dpvc) [reply ts/input/tex/color/ColorMethods.ts:124]: I'm OK either way (probably `''` is better as it is shorter), but would like to see the same treatment both here and at 142, whichever it is.

Comment (pkra) [reply ts/input/tex/color/ColorMethods.ts:124]: Done.

**Comment by dpvc** [comment]
Thanks.  Looks good.

**Pull request merged**